apps/s_socket.c

```diff
@@ -221,6 +221,21 @@ int do_server(int *accept_sock, const char *host, const char *port,
                 break;
             }
             i = (*cb)(sock, type, protocol, context);
+            /*
+             * If we ended with an alert being sent, but still with data in the
+             * network buffer to be read, then calling BIO_closesocket() will
+             * result in a TCP-RST being sent. On some platforms (notably
+             * Windows) then this will result in the peer immediately abandoning
+             * the connection including any buffered alert data before it has
+             * had a chance to be read. Shutting down the sending side first,
+             * and then closing the socket sends TCP-FIN first followed by
+             * TCP-RST. This seems to allow the peer to read the alert data.
+             */
+#ifdef SHUT_WR
```

Comment:
But where does it say that SHUT_WR is universally a macro?

Reply:
Well, if it isn't the worst that happens is that this doesn't get called.

Reply:
Well, it is a macro on Linux but it's also enum, and question is if it's always the case.

Reply:
I don't know...but I'm not sure it matters? Unless you can suggest a better technique for detecting this? The primary motivation for doing this is to fix the tests on Windows (MSYS2). Since the issue, in theory, could be present for any s_server instance talking to a Windows client (maybe other OSs?) then it seems to make sense to fix it for all platforms that we can. But its not a showstopper, so if we're on some platform that we can't reliably detect the presence of this feature - no big deal - we'll just not do it there.

Reply:
> worst that happens is that this doesn't get called.

No, worst is actually SHUT_WR is not macro, while SD_SEND is and totally unrelated (for whichever reason). So that worst is actually shutdown being called with bogus value. Yes, one can argue that it's not a big deal as you are entitled to expect that bogus value will be ignored...

Reply:
Basically, what alerted me here is rather SD_SEND, which is a Windows-specific macro, but is not guarded by _WIN32 condition.

On side note *formally* speaking there should be BIO_shutdown. I mean if we go to the length of providing seemingly "OS-neutral" BIO_closesocket, then there should be neutral BIO_shutdown...

Reply:
So, do you have an alternative suggestion? One possibility is to only do this in Windows (which addresses the test issue without trying to tackle the wider point), e.g.

````
#ifdef (_WIN32)
     shutdown(sock, SD_SEND);
#endif
````

Reply:
> not sure it matters

Neither do I, but it doesn't mean that it's appropriate to not care and not look from other angles.

Reply:
> what alerted me here is rather SD_SEND

I suppose what I'm saying is that it might be more appropriate to
```
#ifdef _WIN32
# ifdef SD_SEND
# endif
#elif defined(SHUT_WR)
#endif
```
At least this is less opportunistic...